File: tutorial/code/py/21-auto-lens.py
from panproto import Panproto, LensHandle, ProtolensChainHandle

panproto = Panproto.init()

# Auto-generate
with LensHandle.auto_generate(old_schema, new_schema, panproto._wasm) as lens:
    view, complement = lens.get(old_data)

# Reusable chain
with ProtolensChainHandle.auto_generate(old_schema, new_schema, panproto._wasm) as chain:
    chain_json = chain.to_json()

# Later, different project
with ProtolensChainHandle.from_json(chain_json, panproto._wasm) as chain2:
    with chain2.instantiate(my_schema) as lens2:
        view, complement = lens2.get(my_data)

File: tutorial/code/ts/21-auto-lens.ts
import { Panproto, LensHandle, ProtolensChainHandle } from '@panproto/core';

const panproto = await Panproto.init();

// Auto-generate a lens between any two schemas
using lens = LensHandle.autoGenerate(oldSchema, newSchema, panproto._wasm);

// The lens works immediately
const { view, complement } = lens.get(oldData);

// Or get the reusable chain for cross-project use
using chain = ProtolensChainHandle.autoGenerate(oldSchema, newSchema, panproto._wasm);

// Save for later — works across schemas with the same structure
const chainJson = chain.toJson();

// Later, in a different project with compatible schemas:
using chain2 = ProtolensChainHandle.fromJson(chainJson, panproto._wasm);
using lens2 = chain2.instantiate(myProjectSchema);
